File: src/main/java/de/wittenbude/expandify/models/db/Episode.java
package de.wittenbude.expandify.models.db;


import de.wittenbude.expandify.models.pojos.Image;
import de.wittenbude.expandify.models.pojos.ResumePoint;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;
import se.michaelthelin.spotify.enums.ModelObjectType;
import se.michaelthelin.spotify.enums.ReleaseDatePrecision;

import java.util.Arrays;
import java.util.Map;

@Data
@NoArgsConstructor
@Document(collection = "episodes")
public class Episode {
    private String audioPreviewUrl;
    private String description;
    private Integer durationMs;
    private Boolean explicit;
    private Map<String, String> externalUrls;
    private String href;

    @Id
    private String id;

    private Image[] images;
    private Boolean isExternallyHosted;
    private Boolean isPlayable;
    private String[] languages;
    private String name;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    private ResumePoint resumePoint;

    @DocumentReference(lazy = true)
    private ShowInfo show;

    private ModelObjectType type;
    private String uri;


    public Episode(se.michaelthelin.spotify.model_objects.specification.Episode episode) {
        this.audioPreviewUrl = episode.getAudioPreviewUrl();
        this.description = episode.getDescription();
        this.durationMs = episode.getDurationMs();
        this.explicit = episode.getExplicit();
        this.externalUrls = episode.getExternalUrls().getExternalUrls();
        this.href = episode.getHref();
        this.id = episode.getId();
        this.images = episode.getImages() == null ? null : Arrays.stream(episode.getImages()).map(Image::new).toArray(Image[]::new);
        this.isExternallyHosted = episode.getExternallyHosted();
        this.isPlayable = episode.getPlayable();
        this.languages = episode.getLanguages();
        this.name = episode.getName();
        this.releaseDate = episode.getReleaseDate();
        this.releaseDatePrecision = episode.getReleaseDatePrecision();
        this.resumePoint = episode.getResumePoint() == null ? null : new ResumePoint(episode.getResumePoint());
        this.show = episode.getShow() == null ? null : new ShowInfo(episode.getShow());
        this.type = episode.getType();
        this.uri = episode.getUri();
    }
}

File: lib/api/schemas/episode.ex
defmodule Episode do
  use Api.Schema

  schema "episodes" do
    field :audio_preview_url, :string
    field :description, :string
    field :html_description, :string
    field :duration_ms, :integer
    field :explicit, :boolean
    embeds_one :external_urls, ExternalUrls
    field :href, :string
    field :spotify_id, :string
    embeds_many :images, Image
    field :is_externally_hosted, :string
    field :is_playable, :boolean
    field :languages, {:array, :string}
    field :name, :string
    field :release_date, :string
    field :release_date_precision, :string
    embeds_one :resume_point, ResumePoint
    field :type, Ecto.Enum, values: [:episode]
    field :uri, :string
    embeds_one :restrictions, Restrictions
    embeds_one :simple_show, EpisodeShow do
      field :description, :string
      field :href, :string
      field :spotify_id, :string
      embeds_many :images, Image
      field :name, :string
      field :uri, :string
      field :total_episodes, :integer
    end

  end

end
